contracts/Proofs.sol

```diff
@@ -121,6 +125,252 @@ contract Proofs {
+  //   x*: blst_fp2
+  //   y*: blst_fp2
+  struct BnP2Affine {
+    BnFp2 x;
+    BnFp2 y;
+  }
+
+  struct BnScalar {
+    // Taken from nim-blscurve/bls_abi, cannot find an analogue in
+    // mratsim/constantine nor nim-bncurve
+    //
+    // # blst_scalar
+    // # = typedesc[array[0..31, byte]]
+    // array[typeof(256)(typeof(256)(256 / typeof(256)(8))), byte]
+    bytes32  ls;
+  }
+
+  struct TauZero {
+    // bytes[512]  name; // array[512, byte], byte is alias for uint8
+    bytes           name;
+    int64           n;
+    Curve.G1Point[] u; // seq[blst_p1]
+  }
+
+  struct Tau {
+    TauZero       t;
+    // bytes32[96] signature;
+    Curve.G1Point signature;
+  }
+
+  // x', y' affine coordinates, result of EllipticCurve.ecMul
+  // e.g. https://github.com/witnet/elliptic-curve-solidity/blob/master/examples/Secp256k1.sol
+  struct PublicKey {
+    Curve.G1Point signkey;
+    Curve.G2Point key;
+  }
+
+  struct QElement {
+    int64     i;
+    BnScalar  v;
+  }
+  function isEmpty(bytes32[96] memory array) internal pure returns (bool) {
+    for(uint i; i< array.length; i++){
+      if(i > 0) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Example of BLS signature verification
+  // Taken from: https://ethereum.stackexchange.com/a/59315
+  function _verifySignature(
+    Curve.G1Point memory sig,
+    Curve.G2Point memory key,
+    uint hashedMsg) internal view returns (bool)
+  {
+    // TODO: Is it ok to use the PublicKey.key (G2) here?
+    // Curve.G2Point memory v = Curve.G2Point(
+    //   [18523194229674161632574346342370534213928970227736813349975332190798837787897, 5725452645840548248571879966249653216818629536104756116202892528545334967238],
+    //   [3816656720215352836236372430537606984911914992659540439626020770732736710924, 677280212051826798882467475639465784259337739185938192379192340908771705870]
+    // );
+
+    Curve.G1Point memory h = Curve.HashToPoint(hashedMsg);
+
+    return Curve.pairingProd2(Curve.g1neg(sig), Curve.P2(), h, key);
+  }
+
+  function _isOnCurve(Curve.G1Point memory g1) internal view returns (bool) {
+    return EllipticCurve.isOnCurve(
+      g1.X,
+      g1.Y,
+      Curve.A(),
+      Curve.B(),
+      Curve.P());
+  }
+  function _verifyProof(
+    Tau memory tau,
+    QElement[] memory q,
+    BnFr[10] memory mus,
+    // Possibly 48 bytes long, csaba?
+    // If only 48 bytes, how can it be a G1Point? (x, y both only 24 bytes?)
+    Curve.G1Point memory sigma,
+    PublicKey memory spk) internal returns (bool) {
+
+    // TODO: is this needed in solidity?
+    // require(!isEmpty(tau.signature), "Signature cannot be empty");
+
+    // TODO: $tau.t -- how to do this in solidity?
+    // TODO: what's the second G2 point needed, PublicKey.signkey is a G1
+    // point?? Is using the PublicKey.key sufficient?
+    uint hashedMsg = uint(sha256(abi.encodePacked(tau.t)));
+    require(_verifySignature(tau.signature, spk.key, hashedMsg),
```

Comment:
I don't think that we need tau.signature. The parameters from the [storage request](https://github.com/status-im/dagger-contracts/blob/2bf01da728317fd1f3ae8424785f31b53eb4013f/contracts/Marketplace.sol#L125) should be enough. When requesting storage, the client submits these parameters in a signed transaction. So no need to verify the signature here.

Reply:
Even though the transaction is signed, how do we know that the signature is valid if we don't verify it?

Reply:
As I understand it, the signature is used to verify that the parameters in Tau (public key, name, etc.) that are presented for verification, are the same ones that were generated by the client. We already know that these parameters are correct, because they are stored in the smart contract after they were submitted by the client in a signed transaction. So tau.signature doesn't need to be generated, stored or verified, if I understand correctly. What do you think @cskiraly?

Reply:
Thanks, Mark. Do we need to check that it's the same signature that's stored in the contract?

Reply:
> Do we need to check that it's the same signature that's stored in the contract?

I think that we don't need the signature at all. We can retrieve the Tau parameters from the contract instead of getting them via the verifyProof parameter, which guarantees that they have been submitted by the client.

Reply:
Thanks guys. Removed.